### ISSUE TITLE
Offline Distillation using Top K Logits

### ISSUE THREAD

**Issue by raghavgarg97** [arcee-ai/DistillKit#18]
Do you guys have any plans to introduce saving top K logits from teacher models and doing KL Training only over those as is mentioned in some of your blogs?

**Comment by fernando-neto-ai** [comment]
Despite being mentioned on the blog, we haven't imported this full part to this opensource pipeline. Though, it is straightforward to implement. We would welcome such contribution

**Comment by raghavgarg97** [comment]
Those looking to implement this, top function that i used are 
```
class LogitsTrainer(SFTTrainer):
    def compute_loss(self, model, inputs, return_outputs=False):
        outputs = model(
            input_ids=inputs['input_ids'],
            attention_mask=inputs['attention_mask'],
            labels=inputs['labels']
        )

        loss = self.distillation_loss(outputs.logits, inputs, outputs.loss)
        return (loss, outputs) if return_outputs else loss

    def distillation_loss(self, student_logits, inputs, original_loss):
        topk_logits = inputs['topk_logits'].to(student_logits.device)
        topk_indices = inputs['topk_indices'].to(student_logits.device)
        attention_mask = inputs['attention_mask'].to(student_logits.device)
        labels = inputs['labels'].to(student_logits.device)

        # Gather student logits corresponding to the teacher's top-k indices
        topk_student_logits = torch.gather(student_logits, dim=-1, index=topk_indices)

        # Apply temperature scaling
        student_logits_scaled = topk_student_logits / distill_temp
        teacher_logits_scaled = topk_logits / distill_temp

        # Compute softmax probabilities
        student_log_probs = F.log_softmax(student_logits_scaled, dim=-1)
        teacher_probs = F.softmax(teacher_logits_scaled, dim=-1)

        # Compute the KL divergence over the top-k logits
        kl_div = F.kl_div(
            student_log_probs,
            teacher_probs,
            reduction='none'
        ).sum(dim=-1)

        # Mask out the prompt tokens (where labels == -100)
        mask = (labels != -100).float()
        kl_div = kl_div * mask

        # Compute the mean KL divergence over the unmasked tokens
        loss_kd = kl_div.sum() / mask.sum()
        loss_kd = loss_kd * (distill_temp ** 2)

        # Combine the original loss with the distillation loss
        loss = distill_alpha * loss_kd + (1 - distill_alpha) * original_loss

        return loss
```
